### PR TITLE
Add zones

### PR DESCRIPTION
**Pull request by retkiewi**
Fixes #42

### CODE REVIEW

File: converter/shieldhit/geo.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+from math import log10, ceil
+
+
+def format_float(number: float, n: int) -> float:
+    """
+    Format float to be up to n characters wide, as precise as possible and as short
+    as possible (in descending priority). so for example given 12.333 for n=5 you will
+    get 12.33, n=7 will be 12.333
+    """
+    if(number == 0):
```
Comment (grzanka) [inline converter/shieldhit/geo.py:13]:
```suggestion
    if number == 0:
```

File: converter/shieldhit/geo.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        return 0.0
+
+    return float(round(number, n-ceil(log10(number))-1))
```
Comment (grzanka) [inline converter/shieldhit/geo.py:16]: Will this work for negative numbers as well, especially `log10`

Comment (grzanka) [reply converter/shieldhit/geo.py:16]: Warn user if number was truncated

File: converter/shieldhit/geo.py [hunk 3]
```diff
@@ -95,18 +95,18 @@ class Zone():
...
+  {id:03d}       {operators:<61}"""
 
     def __str__(self) -> str:
         return self.zone_template.format(
             id=self.id,
-            operators=''.join(['{0:+5}{1:>2}'.format(id, operation)
-                               for id, operation in self.figures_operators]),
+            operators='OR'.join(['  '.join(['{0:+5}'.format(id)
+                                 for id in set])
+                                 for set in self.figures_operators]),
```
Comment (grzanka) [inline converter/shieldhit/geo.py:109]: `set` here shadows basic `set` type

File: converter/shieldhit/geo.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    get 12.33, n=7 will be 12.333
+    """
+    # If number is zero we just want to get 0.0 (it would mess up the log10 operation below)
+    if number == 0:
+        return 0.0
+
+    # Sign messes up the log10 we use do determine how long the number is (we fix that with abs).
+    # Since there is no sign function in python we use `if else`
+    sign = -1 if number < 0 else 1
+    result = float(sign*round(abs(number), n-ceil(log10(number))-1))
```
Comment (grzanka) [inline converter/shieldhit/geo.py:20]: You are still taking `log10(number)` where number may be negative. 
Add good unit tests for negative numbers

File: converter/common.py [hunk 5]
```diff
@@ -14,3 +14,11 @@ def save_configs(self, target_dir: str) -> None:
         Save the configs as text files in the target_dir.
         The files are: beam.dat, mat.dat, detect.dat and geo.dat.
         """
+
+    @abstractmethod
+    def get_configs_json(self) -> dict:
+        """
+        Return a dict representation of the config files.
+        Each file is a field ("beam", "mat", "detect" and "geo") that contains
```
Comment (grzanka) [inline converter/common.py:22]: Python dictionary doesn't have keys, it has list of pairs: key - value.

File: converter/shieldhit/geo.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    """
+    Format float to be up to n characters wide, as precise as possible and as short
+    as possible (in descending priority). so for example given 12.333 for n=5 you will
+    get 12.33, n=7 will be 12.333
+    """
+    # If number is zero we just want to get 0.0 (it would mess up the log10 operation below)
+    if number == 0:
+        return 0.0
+
+    # Adjust n for .
```
Comment (grzanka) [inline converter/shieldhit/geo.py:17]:
```suggestion
    # Adjust n for decimal separator (`.`)
```

File: converter/shieldhit/geo.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    n -= 1
```
Comment (grzanka) [inline converter/shieldhit/geo.py:18]: It is a bad idea to modify parameters, for the reader it is not clear whether the parameter will be modified upon function exit or not, so better to avoid it.
Make a new variable with longer name, assign it to n, then modify accordingly.

File: converter/shieldhit/geo.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+    # Sign messes up the log10 we use do determine how long the number is. We use
+    # abs() to fix that, but we need to remember the sign and update `n` accordingly
+    sign = 1
+
+    if number < 0:
+        number = abs(number)
```
Comment (grzanka) [inline converter/shieldhit/geo.py:25]: Same comment as https://github.com/yaptide/converter/pull/42/files#r755431232

File: converter/shieldhit/geo.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        sign = -1
+        # Adjust n for the sign
+        n -= 1
+
+    whole_length = ceil(log10(number))
+
+    # Check if it will be possible to fit the number
+    if whole_length > n-1:
+        raise ValueError(f"Number is to big to be formatted. Whole part lenght: {whole_length-sign+1},\
```
Comment (grzanka) [inline converter/shieldhit/geo.py:34]: This gives somehow ugly printout:

```
Traceback (most recent call last):
  File "C:\Users\Legion\AppData\Roaming\JetBrains\PyCharm2021.2\scratches\scratch_3.py", line 48, in <module>
    print(format_float(1e12, 8))
  File "C:\Users\Legion\AppData\Roaming\JetBrains\PyCharm2021.2\scratches\scratch_3.py", line 31, in format_float
    raise ValueError(f"Number is to big to be formatted. Whole part lenght: {whole_length-sign+1},\
ValueError: Number is to big to be formatted. Whole part lenght: 12,             requested length: 7
```

File: converter/shieldhit/geo.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+             requested length: {n-sign+1}")
+
+    # Adjust n for the whole numbers, log returns resonable outputs for values greater
```
Comment (grzanka) [inline converter/shieldhit/geo.py:39]:
```suggestion
    # Adjust n for the whole numbers, log returns reasonable outputs for values greater
```

File: converter/shieldhit/geo.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        number = abs(number)
+        sign = -1
+        # Adjust lenght for the sign
+        lenght -= 1
+
+    whole_length = ceil(log10(number))
+
+    # Check if it will be possible to fit the number
+    if whole_length > lenght-1:
+        raise ValueError(f"Number is to big to be formatted. Minimum lenght: {whole_length-sign+1},\
```
Comment (grzanka) [inline converter/shieldhit/geo.py:36]:
```suggestion
        raise ValueError(f"Number is to big to be formatted. Minimum length: {whole_length-sign+1},\
```

File: converter/shieldhit/geo.py [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+             requested length: {n}")
+
+    # Adjust n for the whole numbers, log returns resonable outputs for values greater
+    # than 1, for other values it returns nonpositive numbers, but we would like 1
+    # to be returned. We solve that by taking the greater value between the returned and
+    # and 1.
+    lenght -= max(whole_length, 1)
```
Comment (grzanka) [inline converter/shieldhit/geo.py:43]: Rename to `length`

File: converter/shieldhit/geo.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+    result = float(sign*round(number, lenght))
+
+    # Check if the round function truncated the number, warn the user if it did.
+    if result != sign*number:
```
Comment (grzanka) [inline converter/shieldhit/geo.py:48]: Float comparison is dangerous, in python `0.1 + 0.2` is not equal to `0.3`.
Better to use `math.isclose` with reasonable parameters

File: converter/shieldhit/geo.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+
+#     geo_template: str = """
+# *---><---><--------><------------------------------------------------>
+# {jdbg1:>5}{jdbg1:>5}          {title:>60}
+# *---><---><--------><--------><--------><--------><--------><-------->{figures}
+# END
+# {zones}
+# END
+# {materials}
+# """
```
Comment (grzanka) [inline converter/shieldhit/geo.py:155]: Can it be removed ?

File: converter/shieldhit/parser.py [hunk 15]
```diff
@@ -165,47 +119,91 @@ def save_configs(self, target_dir: str):
...
-            geo_f.write(str(self.geo_config))
+            geo_f.write(self.geo_mat_config.get_geo_string())
+
+    def get_configs_json(self) -> dict:
+        """
+        Return a dict representation of the config files. Each element has
+        the config files name as key and its content as value.
+        """
+        configs_json = {
+            "beam": str(self.beam_config),
```
Comment (grzanka) [inline converter/shieldhit/parser.py:136]: add file extensions as well, this should be proper file names

File: converter/shieldhit/parser.py [hunk 16]
```diff
@@ -165,47 +119,91 @@ def save_configs(self, target_dir: str):
             beam_f.write(str(self.beam_config))
 
         with open(path.join(target_dir, 'mat.dat'), 'x') as mat_f:
-            mat_f.write(str(self.mat_config))
+            mat_f.write(self.geo_mat_config.get_mat_string())
```
Comment (grzanka) [inline converter/shieldhit/parser.py:122]: Use `get_configs_json` to get filenames and file contents.
Putting filenames in multiple places violates DRY principle

File: converter/shieldhit/parser.py [hunk 17]
```diff
@@ -165,47 +119,91 @@ def save_configs(self, target_dir: str):
...
+    def _parse_figures(self, json: dict) -> None:
+        """Parse figures from JSON"""
+        self.geo_mat_config.figures = [solid_figures.parse_figure(
+            figure_dict) for figure_dict in json["scene"]["object"]["children"]]
+
+    def _parse_zones(self, json: dict) -> None:
+        """Parse zones from JSON"""
+        self.geo_mat_config.zones = [
+            Zone(
+                id=(idx+1),
```
Comment (grzanka) [inline converter/shieldhit/parser.py:180]: Do we need brackets ?
Add a comment explaining `+1`

File: tests/test_shieldhit_converter.py [hunk 18]
```diff
@@ -15,7 +16,16 @@
 
 
 _Mat_template = """MEDIUM 0
-ICRU 276
+ICRU 98
```
Comment (grzanka) [inline tests/test_shieldhit_converter.py:19]: I think we need medium numbering from 1

File: tests/test_shieldhit_converter.py [hunk 19]
```diff
@@ -35,23 +45,17 @@
...
-    0    0           protons, H2O 30 cm cylinder, r=10, 1 zone
-*---><---><--------><--------><--------><--------><--------><-------->
-  RCC    1       0.0       0.0       0.0       0.0       0.0      30.0
-                10.0
-  RCC    2       0.0       0.0      -5.0       0.0       0.0      35.0
-                15.0
-  RCC    3       0.0       0.0     -10.0       0.0       0.0      40.0
-                20.0
+    0    0          Unnamed geometry
+  SPH    0       0.0       0.0       0.3       1.0
```
Comment (grzanka) [inline tests/test_shieldhit_converter.py:49]: Should we have numbering from 1 ?

**Review by grzanka — CHANGES_REQUESTED** [review]
see comments

File: converter/shieldhit/parser.py (hunk 15 above)
Comment (retkiewi) [reply converter/shieldhit/parser.py:136]: Should dict keys have '.' in them? That seems unusual to me. Can we swap the '.' for '_' (so it would be mat_dat, geo_dat etc.)?

Comment (grzanka) [reply converter/shieldhit/parser.py]: What can make a problem with `.` ? JS ? From python side its just a normal string

File: tests/test_shieldhit_converter.py [hunk 20]
```diff
@@ -35,23 +45,17 @@
...
+    0    0          Unnamed geometry
+  SPH    1       0.0       0.0       0.3       1.0
+  SPH    2       0.3       0.0       0.0       1.0
+  SPH    3       0.0       0.0      -0.3       1.0
+  SPH    4      -0.3       0.0       0.0       1.0
   END
-  001          +1
-  002          +2     -1
-  003          +3     -2
+  001          +0     -1
```
Comment (grzanka) [inline tests/test_shieldhit_converter.py:54]: Is this line correct ?

**Review by grzanka — APPROVED** [review]
LGTM
